sql: TestCancelDistSQLQuery failed

sql.TestCancelDistSQLQuery [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4189432&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4189432&tab=artifacts#/) on master @ [8cd28089c6c7333615ba3201e841839001d2f0e1](https://github.com/cockroachdb/cockroach/commits/8cd28089c6c7333615ba3201e841839001d2f0e1):


```
=== RUN   TestCancelDistSQLQuery
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestCancelDistSQLQuery2824839686
    test_log_scope.go:80: use -show-logs to present logs inline
    testcluster.go:141: condition failed to evaluate within 3m45s: unexpectedly found 1 active spans:
             0.000ms      0.000ms    === operation:flow _unfinished:1 _verbose:1 node:2
        goroutine 512296 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x72
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x9e917b8, 0xc0053b2000}, 0x653cc00, 0xc00c3a8700)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x7d
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x9e917b8, 0xc0053b2000}, 0xc0053b2000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x5b
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).stopServers(0xc000170b00, {0x9e60860, 0xc000140018})
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:141 +0x55f
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start.func2()
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:356 +0x4b
        github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0xc000625380)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:107 +0x2f
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc0047907e0, {0x9e60860, 0xc000140008})
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:563 +0x391
        github.com/cockroachdb/cockroach/pkg/sql_test.TestCancelDistSQLQuery(0xc0053b2000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/run_control_test.go:128 +0xea0
        testing.tRunner(0xc0053b2000, 0x683c838)
        	/usr/local/go/src/testing/testing.go:1259 +0x230
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1306 +0x727
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestCancelDistSQLQuery2824839686
--- FAIL: TestCancelDistSQLQuery (231.03s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCancelDistSQLQuery.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -402,6 +402,7 @@ func (dsp *DistSQLPlanner) Run(
 	evalCtx *extendedEvalContext,
 	finishedSetupFn func(),
 ) (cleanup func()) {
+	cleanup = func() {}
 	ctx := planCtx.ctx
 
 	flows := plan.GenerateFlowSpecs()
@@ -412,7 +413,7 @@ func (dsp *DistSQLPlanner) Run(
 	}()
 	if _, ok := flows[dsp.gatewayNodeID]; !ok {
 		recv.SetError(errors.Errorf("expected to find gateway flow"))
-		return func() {}
+		return cleanup
 	}
 
 	var (
@@ -464,7 +465,7 @@ func (dsp *DistSQLPlanner) Run(
 		if err != nil {
 			log.Infof(ctx, "%s: %s", clientRejectedMsg, err)
 			recv.SetError(err)
-			return func() {}
+			return cleanup
 		}
 		leafInputState = tis
 	}
@@ -527,9 +528,15 @@ func (dsp *DistSQLPlanner) Run(
 	ctx, flow, opChains, err := dsp.setupFlows(
 		ctx, evalCtx, leafInputState, flows, recv, localState, planCtx.collectExecStats, statementSQL,
 	)
+	// Make sure that the local flow is always cleaned up if it was created.
+	if flow != nil {
+		cleanup = func() {
+			flow.Cleanup(ctx)
+		}
+	}
 	if err != nil {
 		recv.SetError(err)
-		return func() {}
+		return cleanup
 	}
 
 	if finishedSetupFn != nil {
@@ -543,7 +550,7 @@ func (dsp *DistSQLPlanner) Run(
 	if planCtx.saveFlows != nil {
 		if err := planCtx.saveFlows(flows, opChains); err != nil {
 			recv.SetError(err)
-			return func() {}
+			return cleanup
 		}
 	}
 
@@ -556,7 +563,7 @@ func (dsp *DistSQLPlanner) Run(
 	if txn != nil && !localState.MustUseLeafTxn() && flow.ConcurrentTxnUse() {
 		recv.SetError(errors.AssertionFailedf(
 			"unexpected concurrency for a flow that was forced to be planned locally"))
-		return func() {}
+		return cleanup
 	}
 
 	// TODO(radu): this should go through the flow scheduler.
@@ -579,9 +586,7 @@ func (dsp *DistSQLPlanner) Run(
 
 	// ignoreClose is set to true meaning that someone else will handle the
 	// closing of the current plan, so we simply clean up the flow.
-	return func() {
-		flow.Cleanup(ctx)
-	}
+	return cleanup
 }
 
 // DistSQLReceiver is an execinfra.RowReceiver and execinfra.BatchReceiver that

--- a/pkg/sql/flowinfra/flow.go
+++ b/pkg/sql/flowinfra/flow.go
@@ -127,9 +127,11 @@ type Flow interface {
 	// GetID returns the flow ID.
 	GetID() execinfrapb.FlowID
 
-	// Cleanup should be called when the flow completes (after all processors
-	// and mailboxes exited). The implementations must be safe to execute in
-	// case the Flow is never Run() or Start()ed.
+	// Cleanup must be called whenever the flow is done (meaning it either
+	// completes gracefully after all processors and mailboxes exited or an
+	// error is encountered that stops the flow from making progress). The
+	// implementations must be safe to execute in case the Flow is never Run()
+	// or Start()ed.
 	Cleanup(context.Context)
 
 	// ConcurrentTxnUse returns true if multiple processors/operators in the flow
@@ -423,10 +425,7 @@ func (f *FlowBase) IsVectorized() bool {
 
 // Start is part of the Flow interface.
 func (f *FlowBase) Start(ctx context.Context, doneFn func()) error {
-	if err := f.StartInternal(ctx, f.processors, doneFn); err != nil {
-		return err
-	}
-	return nil
+	return f.StartInternal(ctx, f.processors, doneFn)
 }
 
 // Run is part of the Flow interface.

--- a/pkg/sql/flowinfra/flow_scheduler.go
+++ b/pkg/sql/flowinfra/flow_scheduler.go
@@ -28,6 +28,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
 )
 
 const flowDoneChanSize = 8
@@ -177,6 +178,7 @@ func (fs *FlowScheduler) runFlowNow(ctx context.Context, f Flow, locked bool) er
 		fs.mu.Unlock()
 	}
 	if err := f.Start(ctx, func() { fs.flowDoneCh <- f }); err != nil {
+		f.Cleanup(ctx)
 		return err
 	}
 	// TODO(radu): we could replace the WaitGroup with a structure that keeps a
@@ -192,12 +194,13 @@ func (fs *FlowScheduler) runFlowNow(ctx context.Context, f Flow, locked bool) er
 }
 
 // ScheduleFlow is the main interface of the flow scheduler: it runs or enqueues
-// the given flow.
+// the given flow. If the flow is not enqueued, it is guaranteed to be cleaned
+// up when this function returns.
 //
 // If the flow can start immediately, errors encountered when starting the flow
 // are returned. If the flow is enqueued, these error will be later ignored.
 func (fs *FlowScheduler) ScheduleFlow(ctx context.Context, f Flow) error {
-	return fs.stopper.RunTaskWithErr(
+	err := fs.stopper.RunTaskWithErr(
 		ctx, "flowinfra.FlowScheduler: scheduling flow", func(ctx context.Context) error {
 			fs.metrics.FlowsScheduled.Inc(1)
 			telemetry.Inc(sqltelemetry.DistSQLFlowsScheduled)
@@ -217,6 +220,11 @@ func (fs *FlowScheduler) ScheduleFlow(ctx context.Context, f Flow) error {
 			return nil
 
 		})
+	if err != nil && errors.Is(err, stop.ErrUnavailable) {
+		// If the server is quiescing, we have to explicitly clean up the flow.
+		f.Cleanup(ctx)
+	}
+	return err
 }
 
 // NumFlowsInQueue returns the number of flows currently in the queue to be

